[PyTorch] Make RecordFunction store inputs as ArrayRef

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #76266
* #72551
* __->__ #72484

Stepping stone toward stack-allocating array of inputs.

Funnily enough, this seems to improve performance too.

Differential Revision: [D34061169](https://our.internmc.facebook.com/intern/diff/D34061169/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34061169/)!

swolchok committed:
[PyTorch] Make RecordFunction store inputs as ArrayRef

Stepping stone toward stack-allocating array of inputs.

Funnily enough, this seems to improve performance too.

Differential Revision: [D34061169](https://our.internmc.facebook.com/intern/diff/D34061169/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34061169/)!

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/function.h:
```diff
@@ -160,9 +160,10 @@ struct TORCH_API Node : std::enable_shared_from_this<Node> {
           guard.before(
             name(),
-            std::vector<c10::IValue>(inputs.begin(), inputs.end()),
+            c10::ArrayRef<const c10::IValue>(inputs_vec.data(), inputs_vec.size()),
```
Is that ok that the inputs_vec is invalidated just after this call and before anything actually runs?

swolchok replied:
yes -- we debug-check that. before() calls start callbacks, which are the only things that should be touching inputs.

swolchok committed:
Update on "[PyTorch] Make RecordFunction store inputs as ArrayRef"


Stepping stone toward stack-allocating array of inputs.

Funnily enough, this seems to improve performance too.

Differential Revision: [D34061169](https://our.internmc.facebook.com/intern/diff/D34061169/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34061169/)!

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Make RecordFunction store inputs as ArrayRef"


Stepping stone toward stack-allocating array of inputs.

Funnily enough, this seems to improve performance too.

Differential Revision: [D34061169](https://our.internmc.facebook.com/intern/diff/D34061169/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34061169/)!

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Make RecordFunction store inputs as ArrayRef"


Stepping stone toward stack-allocating array of inputs.

Funnily enough, this seems to improve performance too.

Differential Revision: [D34061169](https://our.internmc.facebook.com/intern/diff/D34061169/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34061169/)!

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Make RecordFunction store inputs as ArrayRef"


Stepping stone toward stack-allocating array of inputs.

Funnily enough, this seems to improve performance too.

Differential Revision: [D34061169](https://our.internmc.facebook.com/intern/diff/D34061169/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34061169/)!

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Make RecordFunction store inputs as ArrayRef"


Stepping stone toward stack-allocating array of inputs.

Funnily enough, this seems to improve performance too.

Differential Revision: [D34061169](https://our.internmc.facebook.com/intern/diff/D34061169/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34061169/)!

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Make RecordFunction store inputs as ArrayRef"


Stepping stone toward stack-allocating array of inputs.

Funnily enough, this seems to improve performance too.

Differential Revision: [D34061169](https://our.internmc.facebook.com/intern/diff/D34061169/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34061169/)!

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Make RecordFunction store inputs as ArrayRef"


Stepping stone toward stack-allocating array of inputs.

Funnily enough, this seems to improve performance too.

Differential Revision: [D34061169](https://our.internmc.facebook.com/intern/diff/D34061169/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34061169/)!

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Make RecordFunction store inputs as ArrayRef"


Stepping stone toward stack-allocating array of inputs.

Funnily enough, this seems to improve performance too.

Differential Revision: [D34061169](https://our.internmc.facebook.com/intern/diff/D34061169/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34061169/)!

[ghstack-poisoned]

swolchok commented:
@pytorchbot merge this please

swolchok committed:
Update on "[PyTorch] Make RecordFunction store inputs as ArrayRef"


Stepping stone toward stack-allocating array of inputs.

Funnily enough, this seems to improve performance too.

Differential Revision: [D34061169](https://our.internmc.facebook.com/intern/diff/D34061169/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34061169/)!

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Make RecordFunction store inputs as ArrayRef"


Stepping stone toward stack-allocating array of inputs.

Funnily enough, this seems to improve performance too.

Differential Revision: [D34061169](https://our.internmc.facebook.com/intern/diff/D34061169/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34061169/)!

[ghstack-poisoned]

swolchok commented:
failures look unrelated, landing
